### PR TITLE
SNOW-858328 Add cache for schema evolution permission query

### PR DESCRIPTION
**Pull request by sfc-gh-alhuang**
According to [SNOW-858328](https://snowflakecomputing.atlassian.net/browse/SNOW-858328), add a cache for schema evolution permission query to pull up query to table level and increase the performance.

[SNOW-858328]: https://snowflakecomputing.atlassian.net/browse/SNOW-858328?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:203]:
```suggestion
          tableName2SchemaEvolutionPermission.putIfAbsent(
        tableName,
        this.conn.hasSchemaEvolutionPermission(tableName, connectorConfig.get(SNOWFLAKE_ROLE))
    );
```

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:197]: add a comment here saying what is null
like 
```
null /**schemaEvolutionPermission*/
```

**Review by sfc-gh-japatel — APPROVED** [review]
this is a great improvement, lgtm! Lets get blessing from @sfc-gh-tzhang as well!

**Comment by sfc-gh-japatel** [comment]
Would be great to test it on a db schema with more than one partitions. you can compare two results and see the improvements. 
It would be great if you could also post the latency improvement. Not a blocker but would be good to show.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:192]: @sfc-gh-japatel @sfc-gh-rcheng Is there another place we could put this so that the map doesn't have to be repopulated for every rebalance? Could you move it to the place where we create the client? We should have all the table info from the configuration?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:269]: nit: these three condition checks could be combined?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:198]: Looks like the startTask could be removed now given that we're not using it anymore?

**Review by sfc-gh-tzhang — COMMENTED** [review]
Sorry about the delay, left some comments, PTAL! My main comment is about whether the map could be put in a different place so that there is no need to be repopulated in every rebalance.

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java
Comment (sfc-gh-alhuang) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:221]: Added [here](https://github.com/snowflakedb/snowflake-kafka-connector/blob/e63dbfc17ddf0e045a38b9b19ed6295ce34a7877/src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java#L284).

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java
Comment (sfc-gh-alhuang) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:611]: Removed.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:269]: The boolean expression might looks like this:
```
hasSchemaEvolutionPermission = 
    (hasSchemaEvolutionPermission != null && hasSchemaEvolutionPermission) ||
    (hasSchemaEvolutionPermission == null &&
     this.conn != null && 
     this.conn.hasSchemaEvolutionPermission(tableName, sfConnectorConfig.get(SNOWFLAKE_ROLE))
    );
```
I think it's quit complicate, should I use ternary instead?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:198]: Still using it [here](https://github.com/snowflakedb/snowflake-kafka-connector/blob/e63dbfc17ddf0e045a38b9b19ed6295ce34a7877/src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceFactory.java#L70)

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:192]: The client is lazily created in the [SnowflakeSinkTask](https://github.com/snowflakedb/snowflake-kafka-connector/blob/master/src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java#L158) since we do need the configuration. It would not be affected by rebalance because the [singleton provider](https://github.com/snowflakedb/snowflake-kafka-connector/blob/master/src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java)

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkService.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkService.java:32]: I would not call this `startTasks`

SnowflaekSinkService is in instance associated to the task already. 

Lets call this `startPartitions()`

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:127]: you can drop the javadocs here since it is already present in interface

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:192]: even if we move it to connector level, it is very much possible it needs to repopulated there. that's what is happening with client creation as well.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:269]: IMO it could be combined but it is doesn't read well. current way @sfc-gh-alhuang has is more readable.

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java
Comment (sfc-gh-japatel) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:221]: Nice, thanks @sfc-gh-alhuang

**Review by sfc-gh-japatel — APPROVED** [review]
apart from few naming conventions, LGTM!

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkService.java
Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkService.java]: Fixed.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java
Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java]: Removed.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:193]: Not sure if I understand, so do we have a better place to put this map so that it won't be repopulated every time after rebalance?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:269]: ok, I think the way we're doing this is wried, we should pass a boolean instead of Boolean. It's unexpected that the map doesn't have the table. Then this check could be removed completely, wdyt?

Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java]: @sfc-gh-tzhang make sense. Removed.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:196]: add a single line comment here saying what this false represents

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:125]: we should rename this to `startPartition` instead of `startTask`

Let's create a jira and add a TODO here. You can do this if you have time left after your current PRs are merged.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:198]: I see your concern, valid point. 

I think this can be moved up a level inside a constructor of SnowflakeSinkServiceV2 and we can pass down the map. Is that what you are thinking?

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:198]: What you can do here is call `hasSchemaEvolutionPermission` from inside the Ctor of SnowflakeSinkServiceV2 for every table from topicToTable Map 

pass down the cache in startTask and do a get or if there is cache miss, populate the cache.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java
Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java]: Renamed

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java]: Added.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:200]: what if `tableName2SchemaEvolutionPermission.get(tableName)` is null? have we handled that case?

Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:200]: `createTableIfNotExists` always make sure `tableName` exist. [Ref](https://github.com/snowflakedb/snowflake-kafka-connector/blob/ad5a34544ef706c3e7d9ec2247f057c226312dc6/src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java#L559C2-L559C2)

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:180]: this should also be inside a an uber check of 
`enableSchematization` i.e populate the map only if `enableSchematization` = true

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:130]: why not add the same logic here?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:180]: lets extract this out to a new function which can be called `populateSchemaEvolutionPermissions`

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:200]: I dont see `null` handling. 
So here is what I think we should do:
1. if table is present in cache, do a get. 
2. If not, call permission function and populate it in cache and do a get again. 
3. pass the boolean down from above fetches `createStreamingChannelForTopicPartition`. It is then guaranteed to be non null since your cache is <String, Boolean> and your parameter in function is boolean. 

Let me know if this makes sense.

**Review by sfc-gh-japatel — COMMENTED** [review]
left some comments. PTAL. 
Lets also add a case where in the beginning topic2table has x tables and later in startPartition you add a table which was not part of topic2table. 

Essentially, we are trying to mimick a rebalance case where a table partition was not part of assigned partitions. this will kick in cache miss and we will populate new cache key.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkService.java
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkService.java:23]: thanks for the refactor @sfc-gh-alhuang

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:200]: that check is fine. I am more concerned about the existence of tableName in hashmap. Not in db schema in snowflake. 

there is possibility that a new table/partition gets assigned to a task after rebalance. even though table must have created, your hashmap might not have that tableName.

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:200]: Talked offline, Alec is populating the cache in `createTableIfNotExists`. 
I will wait for the test to be added as mentioned earlier but otherwise looks very promising. Thanks

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java]: Extracted.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:130]: The table2topic is empty in this function.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java]: Added.

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java
Comment (sfc-gh-alhuang) [inline src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java:373]: This test cover `startPartition()` with `topic2Table`

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java
Comment (sfc-gh-japatel) [inline src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java:380]: can you put multiple topics here and then insert records for a topic partition not present in topic2table.

Comment (sfc-gh-alhuang) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java]: Sounds good, added.

**Comment by sfc-gh-japatel** [comment]
LGTM, lets get a signoff from @sfc-gh-tzhang

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:99]: You don't need to initialize here if you choose to do it in the constructor

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:264]: Much better now, thanks for the refactor!

**Review by sfc-gh-tzhang — APPROVED** [review]
LGTM! Left a minor comments, thanks!

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java
Comment (sfc-gh-alhuang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java]: Modified

**Pull request merged**
